src/theory/bv/bvgauss.cpp

```diff
@@ -0,0 +1,715 @@
+/*********************                                                        */
+/*! \file bvgauss.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Aina Niemetz
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2017 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Gaussian Elimination preprocessing pass.
+ **
+ ** Simplify a given equation system modulo a (prime) number via Gaussian
+ ** Elimination if possible.
+ **/
+
+#include "theory/bv/bvgauss.h"
+
+#include <stack>
+#include <unordered_map>
+
+#include "theory/rewriter.h"
+#include "theory/bv/theory_bv_utils.h"
+#include "theory/bv/theory_bv_rewrite_rules_normalization.h"
+
+using namespace CVC4;
+using namespace std;
+
+namespace CVC4 {
+namespace theory {
+namespace bv {
+
+static bool is_bv_const(Node n)
```

Comment:
This method could also return `std::pair<bool, Node>`, then `get_bv_const` would not be needed and additional calls to the rewriter can be avoided. This is probably not a huge performance issue though.

Reply:
I prefer to keep them separate. Additional calls to the rewriter shouldn't make a difference since rewritten terms are cached and IMO using pair here is a bit awkward and does not improve readability.

I think this:
```
Assert(is_bv_const(expr));
...
if (is_bv_const(expr))
{
  expr = get_bv_const(expr);
}
```
is more readable than this:
```
Assert(is_bv_const(expr).first);
...
std::pair<bool, Node> tmp = is_bv_const(expr);
if (tmp.first)
{
  expr = tmp.second;
}
```
Also, node would be copied twice here.

Reply:
In the second example, you wouldn't need the condition. You could just do:

```
bool rewritten;
std::tie(rewritten, expr) = is_bv_const(expr);
```

If you don't care about the `rewritten` bool, you can even do:
```
std::tie(std::ignore, expr) = is_bv_const(expr);
```

But yeah, this is a minor issue, so I won't fight over it ;)

Reply:
No, we need the condition. Since we explicitly do NOT want to rewrite if it is not const.

Reply:
`is_bv_const` can take care of that and return the original node if it does not rewrite to const.

Reply:
Yes, but still, I prefer to keep them separate.